Feature/kakao

BE-11-dongjae opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #114 
- 작업 요약: 소셜 로그인 카카오 기능 개발 했습니다.

## 📄 상세 내용
- [x] 카카오 소셜 로그인 기능 구현 
카카오 OAuth2 인증 절차 처리하는 KakaoService작성

- [x]  serializers 작성 
User 모델의 응답 스펙 정의 UserResponseSerializer 
카카오 로그인 성공 시 응답 스펙 정의 KakaoLoginResponseSerializer

- [x] test code 작성 
code 누락 시 400 응답 
정상 로그인 시 신규 유저 생성 
잘못된 code 또는 API 실패 시 400 응답 
기존 유저일 경우 is_new_user False 인지 확인 
redis 캐시 저장 로직이 호출 되는지 확인 
카카오 id 누락 시 400응답

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

BE-11-dongjae committed:
✨ feat:소셜 로그인 카카오 기능 개발

apps.users.serializers.social_user 작성
apps.users.services.social_user 작성
apps.users.views.social_user 작성
apps.users.test.test_social_user 작성
apps.users.urls.auth_urls에 KakaoLoginCallbackView 연결
config.settings.local에 restAPI, redirect uri 등록

관련 이슈 #114

BE-11-dongjae committed:
🐛 fix : 카카오 로그인 테스트 실패 원인 수정

1) KakaoLoginResponseSerializer user 필드 직렬화 방식 수정
2) views에 UserResponseSerializer 적용
3) test_social_user에 test_cache_set_called 수정

Umdoong commented on .gitignore:
```diff
@@ -10,4 +10,5 @@ htmlcov/
-logs/
\ No newline at end of file
+logs/
+.gitmessage.txt
```
오잉 이거 왜 추가됐나용

BE-11-dongjae replied:
삭제하겠습니다

Umdoong commented on apps/users/serializers/social_user.py:
UserResponseSerializer에 id를 uuid로 바꿔주시고, is_active, is_staff, is_superuser도 추가해주세요

BE-11-dongjae replied:
확인했습니다. 수정하겠습니다.

Meoyoug commented on apps/users/services/social_user.py:
```diff
@@ -0,0 +1,156 @@
+            )
+            return social_user.user, False
+        except SocialUser.DoesNotExist:
+            # 유저를 먼저 생성
```
여기 아래부분은 트랜잭션으로 묶어 보는게 어떨까요?
추가로 User를 생성하기 이전에 만약 같은 이메일이지만 다른 소셜로그인을 진행한 케이스에 대해서도 생각해봐야 할 것 같아요!

BE-11-dongjae replied:
확인했습니다. 수정하고 다시 올리겠습니다

BE-11-dongjae committed:
🐛 fix : 이미 가입된 이메일로 소셜 로그인으로 회원가입 방지
1) services파일에 중복 이메일 가입 방지 로직 추가
2) test에 이메일 중복 방지 테스트 추가
3) test에 반복적으로 사용되는 mock_kakao_api를 셋업데이터에 추가하고 불필요한 중복 줄였습니다.

Umdoong commented on apps/users/views/social_user.py:
```diff
@@ -21,7 +24,17 @@ def get(self, request: Request) -> Response:
+            serializer.is_valid(raise_exception=True)
 
             return Response(serializer.data, status=status.HTTP_200_OK)
 
```
모든 Exception을 400으로 처리하고 있는데, 이러면 나중에 에러를 추적하기 힘들어집니다.
좀 더 구분해서 예외처리를 해주시면 좋을 것 같습니다.

BE-11-dongjae replied:
수정해보겠습니다

Umdoong replied:
이거 Exception 수정 안된 것 같아요~

Umdoong commented on apps/users/serializers/social_user.py:
```diff
@@ -31,4 +31,4 @@ class KakaoLoginResponseSerializer(serializers.Serializer[dict[str, object]]):
     refresh = serializers.CharField()  # JWT 토큰
     is_new_user = serializers.BooleanField()  # 신규유저 여부
-    user = UserResponseSerializer()
+    user = serializers.JSONField()
```
user를 JSONFiled로 해도 되지만, Nasted Serializer를 활용해보시면 더 좋을 것 같습니다.

BE-11-dongjae replied:
찾아보고 수정해보겠습니다

Meoyoug reviewed (changes requested):
로그인 api에서 유저 정보를 모두 건네주는것은 보안상의 이유로 권장되지 않습니다
만약 클라이언트에서 필요하다면 발급받은 엑세스 토큰을 통해 유저의 정보를 조회하는 api를 호출해야합니다.

SimJunSik commented on apps/users/services/social_user.py:
```diff
@@ -0,0 +1,163 @@
+        # 닉네임 중복 방지
+        nickname = base_nickname
+        counter = 1
+        while User.objects.filter(nickname=nickname).exists():
+            nickname = f"{base_nickname}{counter}"
+            counter += 1
```
counter 를 늘릴 때 마다 db query 를 날리는데 개선할 수 있게 한 번 고민해보면 좋을 것 같습니다

BE-11-dongjae committed:
🐛 fix : pr 리뷰 반영

1) UserResponseSerializer 필드 변경 및 추가
2) JSONFiled를 Nasted Serializer로 변경
3) 로그인 api 에서 넘겨주는 유저 정보 최소화
4) 닉네임 중복 방지 로직 수정

Umdoong commented on apps/users/serializers/social_user.py:
```diff
@@ -0,0 +1,48 @@
+    access = serializers.CharField()  # JWT 토큰
+    refresh = serializers.CharField()  # JWT 토큰
+    is_new_user = serializers.BooleanField()  # 신규유저 여부
+    user = UserMinimalResponseSerializer()  # Nested Serializers
```
너무 미니멀한 것 같아요...!! 프론트랑 협의 후 어떤 정보를 넘겨드려야 하는지 얘기해보시는 게 좋을 것 같습니다.

Umdoong commented on config/settings/local.py:
```diff
@@ -19,3 +21,6 @@
     "http://127.0.0.1:3000",
 ]
+
+KAKAO_CLIENT_ID = os.getenv("KAKAO_CLIENT_ID")
```
배포 환경에서도 적용시키려면 base.py나 prod.py에다가 추가하셔야 할 것 같습니다.

Umdoong commented on apps/users/services/social_user.py:
1.카카오에서는 회원가입에 필요한 데이터 중 일부만 가져올 수 있는데, 나머지 데이터들은 어떻게 얻을 건지 생각해보셔야 할 것 같습니다.
2.유저를 만들 때 시리얼라이저를 사용하셔야 합니다.

Umdoong replied:
@BE-11-dongjae 동재님 지금 이게 제일 중요해요!

Umdoong commented:
그리고 커밋 메세지의 경우 PR 반영 이런 것 보단 PR을 반영하여 어떤 것을 수정했는지를 적으시는 게 맞습니다.

BE-11-dongjae committed:
🐛 fix : 예외처리 세분화

BE-11-dongjae committed:
💡 chore : 배포환경에서도 적용시키기 위해 local 내용을 base로 옮김

Umdoong commented:
로컬에서 충돌도 해결해주세요!

BE-11-dongjae committed:
🐛 fix : 테스트코드 추가

BE-11-dongjae committed:
imdaeyong 님이 변경한 내용

BE-11-dongjae committed:
🐛 fix : isort, black 누라
